### ISSUE TITLE
.WMA files along with .XNB files in content/bin

### ISSUE THREAD

**Issue by energyuser** [mono/MonoGame#5002]
Whenever I add a song to my content app, no matter if it's MP3/WAV/WMA I get a .WMA file of the same song together with the XNBs. So I've never published a Monogame project from VS and I guess in the final product there shouldn't be any .WMA files in that folder, right? Otherwise it would be too easy to steal copyrighted songs, for instance. I remember XNA having identical behavior too. So can someone shed some light on this? I don't see .PNG files where my texture XNBs are, neither I see .WAV files where my sounds are... So this only happens with songs. Why?

![Screenshot](http://i.stack.imgur.com/vZDEi.png)

**Comment by KonajuGames** [comment]
PNG is not a native format for consumption by the GPU.  PNG and other image formats are processed to native GPU formats by the content pipeline and the resulting data stored in the XNB file.

With streaming audio, WMA is the native format for consumption by MediaPlayer on Windows.  For Windows platforms, WMA is the selected compressed audio stream format so all input files are converted to WMA by SongProcessor.  The XNB file contains the relative path to the compressed audio file.  The MediaPlayer is given the handle to the file and it takes care of reading the file to play it.

With a bit of extra work, it could be changed to use custom readers and the compressed audio could be included in the XNB file.  It would just be the WMA itself embedded in the file, so extracting it wouldn't be difficult.  It just be a small layer of obscurity.

**Comment by energyuser** [comment]
Extra work on your end? Or I will just have to write a custom content writer?

**Comment by KonajuGames** [comment]
It is extra work in both the content pipeline and especially in the playback code in MonoGame to make it work from a compressed stream embedded in another file.  On some platforms that simply won't work (or at least veru difficult), such as iOS and Android where the OS APIs for playing back compressed audio streams assume separate files.

**Comment by energyuser** [comment]
Well, I'm only developing for Windows and if I were to ever use a song that is not copyright-free it would get saved as a .wma file too...

**Comment by mrhelmut** [comment]
We use licensed music and our contract oblige us to obfuscate song files. To workaround this we are encrypting files in a post-build event, and whenever we load a ```Song```, we decrypt it in the system temp folder and access it from there (and then clean it up when not needed anymore).
We do this only on desktop platforms, since mobiles and consoles are already obfuscated / secure enough.

It's not perfect, I concede, but it's a quick & easy solution which is reliable enough to fully cover our legal requirements (and no sane contract would ask you for an ultimate security, it's mostly always "best effort" because virtually no DRM can be proved secure).

On the other hand, due to this, I've also been looking into embedding Song into XNB. Turns out, as @KonajuGames highlighted, that it's not even possible on some platform and format. For instance, WMA can only be read from the file system on Windows and also requires Windows Media Player to be installed. There is no way to read WMA from streams (on Windows), and it is therefore impossible to embed them in XNBs, unless someone write a WMA decoder for Windows, which sound quite unlikely to happen since Windows has its own perfectly working decoder.
Not to mention mobile platforms, which don't support streaming too.

If you we want MonoGame to support this, one has to come up with a shared format and a stream decoder which works on all platforms. This should also be an option, because it would break XNA compatibility.

**Comment by KonajuGames** [comment]
The other option, which the DesktopGL platforms use, is Ogg Vorbis.  With this, we control the reading of the compressed audio stream and we control the playback.  This is the easy option to get the music embedded in another file or decoded from another stream.

**Comment by energyuser** [comment]
Yeah, I've read somewhere that OGG even has a better compression method than MP3. I'm no audio expert, but I think I'll be using this.
